### PR TITLE
build system: don't optionally use conflicting features

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

As the title says: If a feature is optional and using it would cause a feature conflict, the expectation is that it doesn't end up being used. This adds the required testing.

An additional a unit test to cover this case was added, so that no a future regression would be detected.

### Testing procedure

See https://github.com/RIOT-OS/RIOT/pull/16008#issuecomment-779123619

### Issues/PRs references

Detected in https://github.com/RIOT-OS/RIOT/pull/16008#issuecomment-779123619

### CODE REVIEW

**Comment by fjmolinas** [comment]
Can we really avoid this? Dependency resolution is iterative. If in round one there is only `FEATURES_OPTIONAL=a` it will get used, but if in round `n` you get `FEATURES_BLACKLIST=a` there will still be a conflict. I don't see the harm in the change though, but I don't think it's avoidable, so we can still get this in.

**Comment by maribu** [comment]
> Can we really avoid this? Dependency resolution is iterative. If in round one there is only `FEATURES_OPTIONAL=a` it will get used, but if in round `n` you get `FEATURES_BLACKLIST=a` there will still be a conflict. I don't see the harm in the change though, but I don't think it's avoidable, so we can still get this in.

Both blacklist and list of conflicts should do my understanding not grow. But transitive dependencies to modules might pull in a required feature too late, as an optional feature conflicting with that could be already used.

Still, best effort is likely good enough here for all but a few rare cases. And these could be addressed by tweaking the dependency resolution so that the conflicting required feature gets added in the some round as the optional feature it conflicts with. (But I hope that such occurrences won't show up soon, so that we migrate away in time.)

**Comment by fjmolinas** [comment]
> Both blacklist and list of conflicts should do my understanding not grow. But transitive dependencies to modules might pull in a required feature too late, as an optional feature conflicting with that could be already used.

I think `MODULES` and `PKG` can blacklist things...or maybe it's only TOOLCHAINS? I'm unsure.

**Comment by fjmolinas** [comment]
> Still, best effort is likely good enough here for all but a few rare cases. And these could be addressed by tweaking the dependency resolution so that the conflicting required feature gets added in the some round as the optional feature it conflicts with. (But I hope that such occurrences won't show up soon, so that we migrate away in time.)

Yes, of course, just pointing it out. I doubled checked and mostly its `arch` features that are blacklisted, and those are `FEATURES_REQUIRED`. There is a couple of cases where `bootloader*` and `ble*` are blacklisted as well.

**Comment by fjmolinas** [comment]
> > Still, best effort is likely good enough here for all but a few rare cases. And these could be addressed by tweaking the dependency resolution so that the conflicting required feature gets added in the some round as the optional feature it conflicts with. (But I hope that such occurrences won't show up soon, so that we migrate away in time.)
> 
> Yes, of course, just pointing it out. I doubled checked and mostly its `arch` features that are blacklisted, and those are `FEATURES_REQUIRED`. There is a couple of cases where `bootloader*` and `ble*` are blacklisted as well.

Maybe those can be removed? Maybe just a doc comment? Or we stick with best effort? What do you think?

**Comment by maribu** [comment]
> > Yes, of course, just pointing it out. I doubled checked and mostly its `arch` features that are blacklisted, and those are `FEATURES_REQUIRED`. There is a couple of cases where `bootloader*` and `ble*` are blacklisted as well.
> 
> Maybe those can be removed? Maybe just a doc comment? Or we stick with best effort? What do you think?

I'm not sure if there is indeed so much to gain that it would justify the effort, having the migration to Kconfig in mind. (Just thinking about the amount of quirks in the build system that can be dropped afterwards make me smile :-))

**Comment by fjmolinas** [comment]
Lets spin up the ci

File: makefiles/features_check.inc.mk [hunk 1]
```diff
@@ -1,20 +1,28 @@
 # Check if all required FEATURES are provided
 
+# Features that are used without taking "one out of" dependencies into account
+FEATURES_USED_SO_FAR := $(sort $(FEATURES_REQUIRED) $(FEATURES_OPTIONAL_USED))
+
+_features_would_conflict := $(if $(findstring $1,$2),$(filter-out $1,$(subst :, ,$2)))
+FEATURES_WOULD_CONFLICT := $(foreach features,$(FEATURES_USED_SO_FAR),\
+                             $(call _features_would_conflict,$(features),$(FEATURES_CONFLICT)))
```
Comment (fjmolinas) [inline makefiles/features_check.inc.mk:8]: What is the difference between this and the `_features_conflicting` and `_features_used_conflicting` functions?

File: makefiles/features_check.inc.mk [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
+
 # Features that are provided and not blacklisted
```
Comment (fjmolinas) [inline makefiles/features_check.inc.mk:10]: The comment needs to be fixed.

File: makefiles/features_check.inc.mk [hunk 3]
```diff
@@ -1,20 +1,28 @@
 # Check if all required FEATURES are provided
 
+# Features that are used without taking "one out of" dependencies into account
+FEATURES_USED_SO_FAR := $(sort $(FEATURES_REQUIRED) $(FEATURES_OPTIONAL_USED))
```
Comment (fjmolinas) [inline makefiles/features_check.inc.mk:4]: Why did you move this one? `FEATURES_OPTIONAL_USED` is defined below no?

Comment (maribu) [reply makefiles/features_check.inc.mk:4]: In line 7 this is accessed and needs to be provided by then. `FEATURES_OPTIONAL_USED` on the other hand depends on the usable features, which depends on this.

File: makefiles/features_check.inc.mk (hunk 1 above)
Comment (maribu) [reply makefiles/features_check.inc.mk:8]: These check for potential conflicts in order to avoid them. The existing functions check for actual conflicts that did happen.

The reason why cannot be used for the other is that checking for potential conflicts means for each set of incompatible features you want to filter only those containing a feature that is used and remove the already used features from them. Those are now treated as pseudo-blacklisted, as using them would result in a conflict.

For the other cases one is interested for sets that contain more than one used feature, and you want to print the intersection of the conflicting feature in one set and the actually used features. I just see no way one can be expressed reusing the other. (But that also might be due to either my lack of Makefile-fu or the lack of expressiveness of Makefiles, not by concept.)

File: makefiles/features_check.inc.mk (hunk 3 above)
Comment (fjmolinas) [reply makefiles/features_check.inc.mk:4]: Ah so it catches the previous iteration?

File: makefiles/features_check.inc.mk (hunk 1 above)
Comment (fjmolinas) [reply makefiles/features_check.inc.mk:8]: Ok, no worries, but mind adding some comments to make those functions a bit more understandable on first sight?

Comment (maribu) [reply makefiles/features_check.inc.mk:8]: Oh, there is a logic error in the code. I have to rewrite the function :-(

Comment (maribu) [reply makefiles/features_check.inc.mk]: Fixed and documented

File: makefiles/features_check.inc.mk [hunk 4]
```diff
@@ -3,9 +3,17 @@
...
 
-_features_would_conflict := $(if $(findstring $1,$2),$(filter-out $1,$(subst :, ,$2)))
-FEATURES_WOULD_CONFLICT := $(foreach features,$(FEATURES_USED_SO_FAR),\
-                             $(call _features_would_conflict,$(features),$(FEATURES_CONFLICT)))
+# Get features which inclusion would cause a conflict
+# Parameter 1: Features currently used
+# Parameter 2: A set of features that would conflict (separated by spaces)
+# Algorithm: If interaction of the two lists is empty, return an empty set. Otherwise return
+#            the set of conflicting features without the feature in it that is already used
+_features_would_conflict := $(if $(filter $1,$2),$(filter-out $1,$($2)))
```
Comment (fjmolinas) [inline makefiles/features_check.inc.mk:11]: I don't quite get the `$($2)` here.

Comment (maribu) [reply makefiles/features_check.inc.mk:11]: That's a leftover. I wonder, why this works. Let me double check if the test is working.

Comment (maribu) [reply makefiles/features_check.inc.mk:11]: Aha! Because I placed the unit tests in the wrong folder and it never get's run :/

Comment (maribu) [reply makefiles/features_check.inc.mk]: Btw: Functions (for use with `$(call)` cannot be simply expended variables, the have to be recursively expended variables. Because otherwise they won't store the "function source code", but rather the "functions return value" at the time they are defined.

**Comment by fjmolinas** [comment]
Please squash @maribu

File: makefiles/features_check.inc.mk [hunk 5]
```diff
@@ -1,20 +1,36 @@
...
 
 # Features that may be used, if provided.
 FEATURES_OPTIONAL_ONLY := $(sort $(filter-out $(FEATURES_REQUIRED),$(FEATURES_OPTIONAL)))
 
 # Optional features that end up being used
 FEATURES_OPTIONAL_USED := $(sort $(filter $(FEATURES_USABLE),$(FEATURES_OPTIONAL_ONLY)))
 
+# Update to account for change in FEATURES_OPTIONAL_USED
+FEATURES_USED_SO_FAR := $(sort $(FEATURES_REQUIRED) $(FEATURES_OPTIONAL_USED))
+
```
Comment (fjmolinas) [inline makefiles/features_check.inc.mk:30]: This change seems unneeded now (moving it up)

Comment (maribu) [reply makefiles/features_check.inc.mk]: Moved down again and squashed

**Review by fjmolinas — APPROVED** [review]
ACK!
